### PR TITLE
Official zpt correction

### PR DESCRIPTION
**Pull request by sungbinoh**
Official Zpt correction following the email bellow.
You can call the function in analyzer code as 

vector < Gen > gens = GetGens();
double zpt_reweight = mcCorr->GetOfficialDYReweight(gens);

=============================================
Hi Michael,

The files are available here:

/afs/cern.ch/user/i/ineuteli/public/Zpt/Zpt_weights_201[678].root

You need to compute the Z boson pt and mass at generator-level. As the Z boson (PDG ID 23) is not always explicitly in Drell-Yan events, it's saver to get the two final leptons, e.g. like this:

https://github.com/IzaakWN/NanoTreeProducer/blob/master/corrections/RecoilCorrectionTool.py#L95-L109

Just make sure you get a nice, thin Z mass peak. After that you can get the Z pT weight for any LO MadGraph sample as

https://github.com/IzaakWN/NanoTreeProducer/blob/master/corrections/RecoilCorrectionTool.py#L41-L50

Or if you want to do it on-the-fly with draw, you can use

/afs/cern.ch/user/i/ineuteli/public/Zpt/zptweight.C

E.g.

gROOT.Macro('zptweight.C+')
tree->Draw("pt_1 >> h(100,0,300)","getZpt(m_boson,pt_boson)")

I adopted the method from the HTT group. There is a description in Section 5.6 of AN-19-019_v3. Let me know if you have any other questions.

Cheers,
Izaak
=============================================

### CODE REVIEW

File: AnalyzerTools/src/MCCorrection.C [hunk 1]
```diff
@@ -150,6 +150,11 @@ void MCCorrection::ReadHistograms(){
   }
 */
 
+  // == Get Official DY Pt reweight maps
+  TString DYPtReweightPath = datapath+"/"+TString::Itoa(DataYear,10)+"/DYPtReweight/Zpt_weights_"+TString::Itoa(DataYear,10)+".root";
+  TFile *file_DYPtReweightPath = new TFile(DYPtReweightPath);
+  hist_DYPtReweight_2D = (TH2D *)file_DYPtReweightPath->Get("zptmass_weights");
```
Comment (jedori0228) [inline AnalyzerTools/src/MCCorrection.C]:
```suggestion
  histDir->cd();
  hist_DYPtReweight_2D = (TH2D *)file_DYPtReweightPath->Get("zptmass_weights")->Clone();
  file_DYPtReweightPath->Close();
  delete file_DYPtReweightPath;
  origDir->cd();
```

**Comment by sungbinoh** [comment]
Hist dir lines are added.

File: AnalyzerTools/src/MCCorrection.C [hunk 2]
```diff
@@ -756,3 +764,15 @@ double MCCorrection::GetTopPtReweight(std::vector<Gen> gens){
   return pt_reweight;
 }
 
+double MCCorrection::GetOfficialDYReweight(vector<Gen>& gens){
```
Comment (jedori0228) [inline AnalyzerTools/src/MCCorrection.C]:
```suggestion
double MCCorrection::GetOfficialDYReweight(const vector<Gen>& gens){
```

File: AnalyzerTools/src/MCCorrection.C [hunk 3]
```diff
@@ -756,3 +764,15 @@ double MCCorrection::GetTopPtReweight(std::vector<Gen> gens){
...
+  genFinderDY = new GenFinderForDY();
+  Particle genZ = genFinderDY->Find(gens);
+  mZ = genZ.M();
```
Comment (jedori0228) [inline AnalyzerTools/src/MCCorrection.C]:
```suggestion
  double mZ = genZ.M();
```

File: AnalyzerTools/src/MCCorrection.C [hunk 4]
```diff
@@ -756,3 +764,15 @@ double MCCorrection::GetTopPtReweight(std::vector<Gen> gens){
...
+  ptZ = genZ.Pt();
```
Comment (jedori0228) [inline AnalyzerTools/src/MCCorrection.C]:
```suggestion
  double ptZ = genZ.Pt();
```

File: AnalyzerTools/src/MCCorrection.C (hunk 2 above)
Comment (jedori0228) [reply AnalyzerTools/src/MCCorrection.C]: and also header

File: AnalyzerTools/include/GenFinderForDY.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+class GenFinderForDY{
+
+public:
+
+  GenFinderForDY();
+  ~GenFinderForDY();
+
+  bool FoundGenZ;
+  Particle Find(vector<Gen>& gens);
```
Comment (jedori0228) [inline AnalyzerTools/include/GenFinderForDY.h]:
```suggestion
  Particle Find(const vector<Gen>& gens);
```

**Comment by jedori0228** [comment]
I fixed couple of things including bugs above in my repo.
Please merge jedori0228:OfficialZptCorrection into your sungbinoh:Official_Zpt_Correction branch
To get the reweight, follow below lines in your analyzer;

  //==== Reweight
  double ZPtReweight = mcCorr->GetOfficialDYReweight(gens);
  //==== Access to the genZ (<- some updates are done to make this available)
  Particle genz = mcCorr->genFinderDY->GenZ;
  FillHist("M", genz.M(), 1., 1000, 0., 1000.);
  FillHist("Pt", genz.Pt(), 1., 1000, 0., 1000.);

**Pull request merged**